### PR TITLE
added optional rounding solver options

### PR DESCRIPTION
**Pull request by wrangelvid**
In practice, it is sufficient to solve the relaxed `GraphOfConvexSets` problem with lower tolerances and with higher tolerances for the simpler rounded problems where the flows are set. This dramatically reduces the overall solve time.

- - -
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18875)

### CODE REVIEW

**Review by mpetersen94 — COMMENTED** [review]
+@mpetersen94 for feature review

Running `/usr/bin/clang-format-12 -style=file -i bindings/pydrake/geometry/geometry_py_optimization.cc` should fix the CI issues.

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18875#-:-NPERhn-EUuY8-1kXA2W:b5voxa8)* status: 3 unresolved discussions, LGTM missing from assignee mpetersen94, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @wrangelvid)

___
*[`geometry/optimization/graph_of_convex_sets.cc` line 631 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18875#-NPEUroFE62tFZDiV4n8:-NPEUroFE62tFZDiV4n9:b61upvr) ([raw file](https://github.com/RobotLocomotion/drake/blob/21e2a9fb321a1aa0e382ef5096b7dd7c36501860/geometry/optimization/graph_of_convex_sets.cc#L631)):*
> ```cpp
>     options.solver->Solve(
>         prog, {},
>         (rounding && options.rounding_solver_options != std::nullopt)
> ```

Here and below you can achieve the same results with:
`rounding && options.rounding_solver_options`

Also, this inline if/else is hard to parse. Do you think it could be pulled out of the Solve argument? Instead the right solver options to pass could be figured out once (i.e. right after the result is declared).
___
*[`geometry/optimization/test/graph_of_convex_sets_test.cc` line 1323 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18875#-NPEXUtfA6w1dlHoKjF8:-NPEXUtfA6w1dlHoKjF9:b1pfjwo) ([raw file](https://github.com/RobotLocomotion/drake/blob/21e2a9fb321a1aa0e382ef5096b7dd7c36501860/geometry/optimization/test/graph_of_convex_sets_test.cc#L1323)):*
> ```cpp
>   options.preprocessing = true;
>   options.max_rounded_paths = 10;
>   options.rounding_solver_options = SolverOptions();
> ```

This is not a sufficient test of the new API. Making the change to the .h file and making no change to the .cc file would still allow this test to pass.  If you want help figuring out a good test of the new API let's talk offline.
___
*[`bindings/pydrake/geometry/geometry_py_optimization.cc` line 390 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18875#-NPEaaJ_20LqqNMpR4ze:-NPEaaJ_20LqqNMpR4zf:b6od08a) ([raw file](https://github.com/RobotLocomotion/drake/blob/21e2a9fb321a1aa0e382ef5096b7dd7c36501860/bindings/pydrake/geometry/geometry_py_optimization.cc#L390)):*
> ```cpp
>             }),
>             cls_doc.solver_options.doc)
>         .def_property("rounding_solver_options",
> ```

This also needs a test to confirm that the bindings are correct and enable the desired read/write behavior.

**Review by wrangelvid — COMMENTED** [review]
Done.

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18875#-:-NPLKrPmDXJLifFNkV9o:b-896fix)* status: 3 unresolved discussions, LGTM missing from assignee mpetersen94, needs platform reviewer assigned, needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @mpetersen94)

___
*[`geometry/optimization/graph_of_convex_sets.cc` line 631 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18875#-NPEUroFE62tFZDiV4n8:-NPLL4Z56QUYmsdtdAL4:b-896fix) ([raw file](https://github.com/RobotLocomotion/drake/blob/21e2a9fb321a1aa0e382ef5096b7dd7c36501860/geometry/optimization/graph_of_convex_sets.cc#L631)):*
<details><summary><i>Previously, mpetersen94 (Mark Petersen) wrote…</i></summary><blockquote>

Here and below you can achieve the same results with:
`rounding && options.rounding_solver_options`

Also, this inline if/else is hard to parse. Do you think it could be pulled out of the Solve argument? Instead the right solver options to pass could be figured out once (i.e. right after the result is declared).
</blockquote></details>

Done.
___
*[`geometry/optimization/test/graph_of_convex_sets_test.cc` line 1323 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18875#-NPEXUtfA6w1dlHoKjF8:-NPLL6UEAL6UlP2n5Zt3:b-896fix) ([raw file](https://github.com/RobotLocomotion/drake/blob/21e2a9fb321a1aa0e382ef5096b7dd7c36501860/geometry/optimization/test/graph_of_convex_sets_test.cc#L1323)):*
<details><summary><i>Previously, mpetersen94 (Mark Petersen) wrote…</i></summary><blockquote>

This is not a sufficient test of the new API. Making the change to the .h file and making no change to the .cc file would still allow this test to pass.  If you want help figuring out a good test of the new API let's talk offline.
</blockquote></details>

Done.
___
*[`bindings/pydrake/geometry/geometry_py_optimization.cc` line 390 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18875#-NPEaaJ_20LqqNMpR4ze:-NPLLB1M5eZsVU4sRypT:b-896fix) ([raw file](https://github.com/RobotLocomotion/drake/blob/21e2a9fb321a1aa0e382ef5096b7dd7c36501860/bindings/pydrake/geometry/geometry_py_optimization.cc#L390)):*
<details><summary><i>Previously, mpetersen94 (Mark Petersen) wrote…</i></summary><blockquote>

This also needs a test to confirm that the bindings are correct and enable the desired read/write behavior.
</blockquote></details>

Done.

**Review by wrangelvid — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18875#-:-NTfTisz2PjmH31C35IH:bnfp4nl)* status: 3 unresolved discussions, LGTM missing from assignee mpetersen94, needs platform reviewer assigned, needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @mpetersen94)

**Review by RussTedrake — COMMENTED** [review]
-@mpetersen94 
+@RussTedrake for feature review.
"lgtm:

> Reviewed 1 of 4 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18875#-:-NTfdXbp7P-TRrzUDaCC:b-35mdwh)* status: 3 unresolved discussions, LGTM missing from assignee RussTedrake(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @mpetersen94)

**Review by RussTedrake — COMMENTED** [review]
+@ggould-tri for platform review, please. per the schedule.

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18875#-:-NTfi-1l0tQmHyaNN0nH:bxbqbap)* status: 3 unresolved discussions, LGTM missing from assignees RussTedrake(platform),ggould-tri(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @ggould-tri and @mpetersen94)

**Review by mpetersen94 — COMMENTED** [review]
:lgtm:

> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18875#-:-NTl818zEoRUX-Ot1hEr:bnfp4nl)* status: LGTM missing from assignees ggould-tri(platform),RussTedrake(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @ggould-tri)

**Review by ggould-tri — COMMENTED** [review]
:lgtm: with one clarification request on the documentation.

> Reviewed 1 of 4 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18875#-:-NTnbXWv6_0W5yah8DgS:b8flahe)* status: 2 unresolved discussions, LGTM missing from assignee RussTedrake(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @wrangelvid)

___
*[`geometry/optimization/graph_of_convex_sets.h` line 72 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/18875#-NTn_ZgB9kF_b6kLETHc:-NTn_ZgB9kF_b6kLETHd:bgbcvd7) ([raw file](https://github.com/RobotLocomotion/drake/blob/2be3ab9b9ace52d467bbf01fbd6ee8d4c45cebdb/geometry/optimization/graph_of_convex_sets.h#L72)):*
> ```objc
>   If not set, solver_options is used.
>   For instance, one might want to set lower tolerances for running
>   the relaxed problem and higher tolerances for final solves during rounding. */
> ```

minor:  I found "higher" and "lower" here a bit confusing, particularly in the context of "relaxed" -- I'm not quite sure from this comment which one should be the big number and which one is the small number, and the tests don't provide an example to crib from.
Using the terms "tighter" and "looser" might more clearly distinguish numerically larger, looser tolerances from numerically smaller, tighter tolerances.
For instance, assuming that I am reading the intent of this comment correctly:



_Suggestion:_
```objc
  For instance, one might want to set tighter (i.e., lower) tolerances for running
  the relaxed problem and looser (i.e., higher) tolerances for final solves during rounding. */
```
___
*[`geometry/optimization/test/graph_of_convex_sets_test.cc` line 1382 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/18875#-NTnbHq3-Bb4GgMpdYja:-NTnbHq3-Bb4GgMpdYjb:bb7phvt) ([raw file](https://github.com/RobotLocomotion/drake/blob/2be3ab9b9ace52d467bbf01fbd6ee8d4c45cebdb/geometry/optimization/test/graph_of_convex_sets_test.cc#L1382)):*
> ```cpp
> 
>     // Without the convex relaxation, the solver should ignore the
>     // rounded_solver_options and succeed.
> ```

typo



_Suggestion:_
```cpp
rounding_solver_options
```

**Review by ggould-tri — COMMENTED** [review]
(FYI, @RussTedrake , I believe your 'lgtm' emoji wasn't formatted correctly and was not noted by reviewable.)

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18875#-:-NTnblfJB2ZGB731GSwo:bkz0v94)* status: 2 unresolved discussions, LGTM missing from assignee RussTedrake(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @wrangelvid)

**Review by wrangelvid — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18875)* status: LGTM missing from assignee RussTedrake(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @wrangelvid)

___
*[`geometry/optimization/graph_of_convex_sets.h` line 72 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/18875#-NTn_ZgB9kF_b6kLETHc:-NTnehPgCky34BCbBbA5:b-lv7iq9) ([raw file](https://github.com/RobotLocomotion/drake/blob/2be3ab9b9ace52d467bbf01fbd6ee8d4c45cebdb/geometry/optimization/graph_of_convex_sets.h#L72)):*
> For instance, one might want to set tighter (i.e., lower) tolerances for running
>   the relaxed problem and looser (i.e., higher) tolerances for final solves during rounding. */
Done.
___
*[`geometry/optimization/test/graph_of_convex_sets_test.cc` line 1382 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/18875#-NTnbHq3-Bb4GgMpdYja:-NTnejjR0FFCGDfJaLVY:b-896fix) ([raw file](https://github.com/RobotLocomotion/drake/blob/2be3ab9b9ace52d467bbf01fbd6ee8d4c45cebdb/geometry/optimization/test/graph_of_convex_sets_test.cc#L1382)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

typo
</blockquote></details>

Done.

**Comment by RussTedrake** [comment]
:lgtm:

**Review by RussTedrake — COMMENTED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18875)* status: commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @wrangelvid)
